spring-data-jdbc/src/main/java/org/springframework/data/jdbc/core/convert/BasicJdbcConverter.java

```diff
@@ -129,5 +161,65 @@ public Object writeValue(@Nullable Object value, TypeInformation<?> type) {
 		return super.writeValue(value, type);
 	}
 
+	@Override
+	public boolean canWriteValue(@Nullable Object value, TypeInformation<?> type) {
 
+		if (value == null) {
+			return true;
+		}
+
+		if (AggregateReference.class.isAssignableFrom(value.getClass())) {
+			return canWriteValue(((AggregateReference) value).getId(), type);
+		}
+		return super.canWriteValue(value, type);
+	}
+
+	public JdbcTypeAware writeTypeAware(@Nullable Object value, Class<?> columnType, int sqlType) {
+
+		JdbcTypeAware jdbcTypeAware = tryToConvertToJdbcTypeAware(value);
+		if (jdbcTypeAware != null) {
+			return jdbcTypeAware;
+		}
+
+		Object convertedValue = writeValue(value, ClassTypeInformation.from(columnType));
+
+		if (convertedValue == null || !convertedValue.getClass().isArray()) {
+			return JdbcTypeAware.of(convertedValue, JdbcUtil.jdbcTypeFor(sqlType));
+		}
+
+		Class<?> componentType = convertedValue.getClass().getComponentType();
+		if (componentType != byte.class && componentType != Byte.class) {
+			return JdbcTypeAware.of(typeFactory.createArray((Object[]) convertedValue), JDBCType.ARRAY);
+		}
+
+		if (componentType == Byte.class) {
+			Byte[] boxedBytes = (Byte[]) convertedValue;
```

Comment:
Let's extract the code into a package-private utility (`ArrayUtil.toPrimitive(…)`) or something like that.